packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -356,6 +373,30 @@ export class PendingStateManager implements IDisposable {
         this.isProcessingBatch = false;
     }
 
+    /**
+     * Capture the pending state at this point
+     */
+    public checkpoint() {
+        const checkpointHead = this.pendingStates.peekBack();
+        return {
+            rollback:() => {
+                try {
+                    while(this.pendingStates.peekBack() !== checkpointHead) {
+                        this.rollbackNextPendingState();
+                    }
+                } catch(err) {
+                    const error = DataProcessingError.wrapIfUnrecognized(
```

Comment:
I think you may just want to use `wrapError` directly - the way you have to plumb through the message prefix to get the result you want is ending up breaking the intended semantics of `wrapIfUnrecognized` and `normalizeError`.

I think that makes sense for this codepath anyway - You want to return a DataProcessingError no matter what here.